subsys/mgmt/mcumgr/transport/smp_bt.c

```diff
@@ -502,41 +551,84 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 {
 	struct conn_param_data *cpd = conn_param_data_get(conn);
 
-	/* Cancel work if ongoing. */
-	(void)k_work_cancel_delayable(&cpd->dwork);
-	(void)k_work_cancel_delayable(&cpd->ework);
+	/* Clear all pending requests from this device which have yet to be processed from the
+	 * FIFO (for this specific connection).
+	 */
+	smp_rx_clear(&smp_bt_transport, false, (void *)conn);
+
+	/* Force giving the notification semaphore here, this is only needed if there is a pending
+	 * outgoing packet when the device has disconnected, as in this case the notification
+	 * callback will not be called and this is needed to prevent a deadlock.
+	 */
+	k_sem_give(&cpd->smp_notify_sem);
+
+	if (IS_ENABLED(CONFIG_MCUMGR_SMP_BT_CONN_PARAM_CONTROL)) {
+		/* Cancel work if ongoing. */
+		(void)k_work_cancel_delayable(&cpd->dwork);
+		(void)k_work_cancel_delayable(&cpd->ework);
+
+		/* Clear cpd. */
+		cpd->state = 0;
+	}
 
 	/* Clear cpd. */
-	cpd->state = 0;
+	cpd->id = 0;
 	cpd->conn = NULL;
 }
 
 static void conn_param_control_init(void)
 {
-	/* Register BT callbacks */
-	static struct bt_conn_cb conn_callbacks = {
-		.connected = connected,
-		.disconnected = disconnected,
-	};
-	bt_conn_cb_register(&conn_callbacks);
-
 	for (size_t i = 0; i < ARRAY_SIZE(conn_data); i++) {
 		k_work_init_delayable(&conn_data[i].dwork, conn_param_on_pref_restore);
 		k_work_init_delayable(&conn_data[i].ework, conn_param_on_error_retry);
 	}
 }
 
+static bool smp_bt_clear_check(struct net_buf *nb, void *arg)
+{
+	const struct bt_conn *conn = (struct bt_conn *)arg;
+	struct smp_bt_user_data *ud = net_buf_user_data(nb);
+	struct conn_param_data *cpd;
+
+	if (conn == NULL || ud == NULL) {
+		return true;
+	}
+
+	cpd = conn_param_data_get(conn);
+
+	if (ud->conn == conn && (cpd->id == 0 || cpd->id == ud->id)) {
```

Comment:
If I understand correctly, this is checking that the smp buffer belongs to the conn obj passed in `arg`, right?
Why do we need the `cpd->id == 0` part?

edit: I see this block in `connected`
```c
			if (cpd->id == 0) {
				++cpd->id;
				++next_id;
			}
```
does that mean that `id==0` has a special meaning?

Reply:
ID = 0 means unused, ID != 0 means used